File: progressive_tax_calculator/progressive_tax_calculator.go
package progressive_tax_calculator

import (
	"github.com/mieubrisse/financial-predictor/tax_bracket_parser"
	"github.com/sirupsen/logrus"
	"math"
	"sort"
)

func Calculate(brackets []tax_bracket_parser.TaxBracket, amount float64) float64 {
	// Sort the brackets in REVERSE order
	comesBeforeFunc := func(firstElemIdx int, secondElemIdx int) bool {
		return brackets[firstElemIdx].Floor > brackets[secondElemIdx].Floor
	}
	sort.Slice(brackets, comesBeforeFunc)

	var totalTax float64 = 0
	for _, bracket := range(brackets) {
		floor := bracket.Floor
		rate := bracket.Rate
		logrus.Debugf("Floor: %v, Rate: %v", floor, rate)
		amountInBracket := math.Max(0, amount - floor)
		totalTax += rate * amountInBracket
		amount -= amountInBracket
		logrus.Debugf("Tax so far: %v", totalTax)
	}
	return totalTax
}

File: src/main/java/com/strangegrotto/wealthdraft/tax/ProgressiveTaxCalculator.java
package com.strangegrotto.wealthdraft.tax;

import com.google.common.collect.Ordering;
import com.strangegrotto.wealthdraft.govconstants.TaxBracket;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

import java.util.ArrayList;
import java.util.Comparator;
import java.util.List;

public class ProgressiveTaxCalculator {
    private static final Logger LOG = LoggerFactory.getLogger(ProgressiveTaxCalculator.class);

    private final List<TaxBracket> brackets;

    public ProgressiveTaxCalculator(List<TaxBracket> brackets) {
        List<TaxBracket> bracketsCopy = new ArrayList(brackets);

        // Store the brackets in REVERSE order, since we need it for calculating
        Comparator<TaxBracket> bracketComparator = Comparator.comparing(
                bracket -> bracket.getFloor()
        );
        Ordering<TaxBracket> taxBracketOrdering = Ordering.from(bracketComparator)
                .reverse();
        bracketsCopy.sort(taxBracketOrdering);
        this.brackets = bracketsCopy;
    }

    // TODO switch from using double (which has precision errors) to using BigDecimal or something similar
    public double calculateTax(long amount) {
        long remainingToCalculateTaxFor = amount;
        double totalTax = 0;
        for (TaxBracket bracket : this.brackets) {
            long floor = bracket.getFloor();
            double rate = bracket.getRate();
            LOG.debug("Floor: {}, Rate: {}", floor, rate);

            long amountInBracket = Math.max(0, remainingToCalculateTaxFor - floor);
            double taxInBracket = (double)amountInBracket * rate;
            totalTax += taxInBracket;
            remainingToCalculateTaxFor -= amountInBracket;

            LOG.debug("Tax so far: {}", totalTax);
        }
        return totalTax;
    }
}
